rpcutils/rpcclient.go

```diff
@@ -631,3 +633,185 @@ func APITransaction(client txhelpers.VerboseTransactionGetter, txid *chainhash.H
+			checkOutputs = true
+			break
+		}
+	}
+	if checkOutputs {
+		contracts, err := checkTxOutputsForSwapInfo(client, txraw, spendingTxsFn, params)
+		if err != nil {
+			return nil, err
+		}
+		if len(contracts) > 0 {
+			tx.Contracts = contracts
+			appendFound("Contract")
+		}
+	}
+
+	// Check which tx inputs spend the output of an atomic swap contract.
+	for i, input := range txraw.Vin {
+		inputIndex := uint32(i)
+		swapInfo, err := checkTxInputForSwapInfo(txid, inputIndex, input, params)
+		if err != nil {
+			return nil, fmt.Errorf("error checking if input redeems a contract: %v", err)
+		}
+		if swapInfo == nil {
+			continue
+		}
+		if swapInfo.RedeemedBy == txhelpers.AtomicSwapInitiator {
+			tx.Refunds[inputIndex] = swapInfo
+			appendFound("Refund")
+		} else {
+			tx.Redemptions[inputIndex] = swapInfo
+			appendFound("Redemption")
+		}
+	}
+
+	return tx, nil
+}
+
+func checkTxOutputsForSwapInfo(client txhelpers.VerboseTransactionGetter, tx *chainjson.TxRawResult,
+	spendingTxsFn SpendingTxsFn, params *chaincfg.Params) (map[uint32]*apitypes.AtomicSwap, error) {
+
+	spendingTxHashes, spendingTxVinInds, spentVoutInds, err := spendingTxsFn(tx.Txid)
+	if err != nil {
+		return nil, fmt.Errorf("error getting outputs spending info: %v", err)
+	}
+
+	// validate the returned spending info before making rpc calls to retrieve
+	// spending inputs
+	if len(spendingTxHashes) > len(tx.Vout) {
+		return nil, fmt.Errorf("invalid spending transaction data for %s", tx.Txid)
+	}
+	for _, voutIndex := range spentVoutInds {
+		if int(voutIndex) >= len(tx.Vout) {
+			return nil, fmt.Errorf("invalid spending transaction data for %s", tx.Txid)
+		}
+	}
+
+	// Check which of the spent outputs is a contract.
+	contracts := make(map[uint32]*apitypes.AtomicSwap)
+	for i, voutIndex := range spentVoutInds {
+		vout := tx.Vout[voutIndex]
+		if vout.ScriptPubKey.Type != txscript.ScriptHashTy.String() {
+			// ignore spent non-p2sh outputs
+			continue
+		}
+		spendingTxHash, spendingTxIndex := spendingTxHashes[i], spendingTxVinInds[i]
+		spendingInput, err := TxInput(client, spendingTxHash, spendingTxIndex)
+		if err != nil {
+			return nil, fmt.Errorf("error finding input that spends txout %s:%d at %s:%d -> %v",
+				tx.Txid, voutIndex, spendingTxHash, spendingTxIndex, err)
+		}
+		swapInfo, err := checkTxInputForSwapInfo(spendingTxHash, spendingTxIndex, *spendingInput, params)
+		if err != nil {
+			return nil, fmt.Errorf("error checking if tx output is a contract: %v", err)
+		}
+		if swapInfo != nil {
+			contracts[voutIndex] = swapInfo
+		}
+	}
+	return contracts, nil
+}
+
+// checkTxInputForSwapInfo parses the scriptsig of the provided transaction input
+// for information about a completed atomic swap.
+// Returns (nil, nil) if the scriptsig of the provided txin does not redeem a
+// swap contract.
+func checkTxInputForSwapInfo(txid string, inputIndex uint32, input chainjson.Vin, params *chaincfg.Params) (*apitypes.AtomicSwap, error) {
```

Comment:
This would be a better fit for `txhelpers`.  One of the functions I'm alluding to in my above review comment.